velox/expression/EvalCtx.cpp

```diff
@@ -273,6 +273,49 @@ VectorPtr EvalCtx::ensureFieldLoaded(
   return field;
 }
 
+namespace {
+// Helper function for addNulls that ensures result is unique and have writable
+// nulls for size.
+// The method should also guarantee that result is not constant since addNulls
+// is not defined for constant vectors.
+void prepareForNullWriting(
+    VectorPtr& result,
+    vector_size_t size,
+    EvalCtx& context) {
+  if (result->isFlatEncoding()) {
+    if (!result.unique() || !result->isNullsWritable()) {
+      BaseVector::ensureWritable(
+          SelectivityVector::empty(), result->type(), context.pool(), result);
+    }
+    if (result->size() < size) {
+      result->resize(size);
+    }
+    return;
+  }
+
+  // Complex types or non-flat encodings, we wrap result in dictionary vector.
+  auto dictionarySize = std::max(size, result->size());
+  auto indices = allocateIndices(dictionarySize, context.pool());
+
+  auto* rawIndices = indices->asMutable<vector_size_t>();
+  for (int i = 0; i < result->size(); i++) {
+    rawIndices[i] = i;
+  }
+
+  BufferPtr nulls = nullptr;
+  if (result->size() < size) {
+    nulls = allocateNulls(dictionarySize, context.pool());
+    auto* mutableNulls = nulls->asMutable<uint64_t>();
+    for (int i = result->size(); i < size; ++i) {
+      bits::setNull(mutableNulls, i, true);
+    }
+  }
+
+  result = BaseVector::wrapInDictionary(
```

Comment:
Should we check if result is a unique null constant, we just need to resize it?

Reply:
this is already checked before callign this function